dvc/fs/repo.py

```diff
@@ -340,18 +329,34 @@ def walk(
                 onerror(NotADirectoryError(top))
             return
 
-        ignore_subrepos = not self._traverse_subrepos
-        if follow_subrepos is not None:
-            ignore_subrepos = not follow_subrepos
+        repo = self._get_repo(os.path.abspath(top))
+        ignore_subrepos = kwargs.pop("ignore_subrepos", True)
+        if repo:
+            yield from repo.dvcignore(
+                self.walk_fs(
+                    top,
+                    topdown=topdown,
+                    onerror=onerror,
+                    dvcfiles=dvcfiles,
+                    **kwargs
+                ),
+                ignore_subrepos=ignore_subrepos,
+            )
+        else:
+            yield from self.walk_fs(
+                top,
+                topdown=topdown,
+                onerror=onerror,
+                dvcfiles=dvcfiles,
+                **kwargs
+            )
```

Comment:
Duplicated code :slightly_frowning_face: Seems like dvcignore noop would come in handy in cleaning it up.

I also see that the approach with `dvcignore.walk(fs, path_info)` from above will run into a problem of `walk_fs` having a custom name, so you probably need to apply dvcignore inside `walk_fs`, for fs/dvcfs walks.

Reply:
For the record: that existing walk logic for fs & dvc_fs is quite messy, we are aware of that, so just take a look if it is possible to wrap with dvcignore without messing with it too much. The code itself has been due for a refactor for awhile now :slightly_frowning_face:

Reply:
> I also see that the approach with dvcignore.walk(fs, path_info) from above will run into a problem of walk_fs having a custom name, so you probably need to apply dvcignore inside walk_fs, for fs/dvcfs walks.

Sorry, I don't get it. Currently, the `dvcignore` can work correctly in `subrepo`s. We don't need to pass `fs` arguments to it.

Reply:
@karajan1001 Actually not really. There is no need to wrap `self._walk`, for example. You just need to wrap `fs.walk`, as `dvc_fs.walk` already used dvcignore when it built the dag, so there is no reason to apply it again.

Reply:
Also, if we'll be able to pull that off ^^ , then we could switch to:

```
dvcignore.walk(fs, path)
dvcignore.walk_files(fs, path)
```

instead of current

```
dvcignore.walk(fs.walk(path))
dvcignore.walk_files(fs.walk(path))
```

Unless I'm missing something, of course.